libraries/standard/mqtt/src/mqtt.c

```diff
@@ -647,18 +729,44 @@ static MQTTStatus_t handleIncomingPublish( MQTTContext_t * pContext,
                                           MQTT_RECEIVE,
                                           publishInfo.qos,
                                           &publishRecordState );
-        LogInfo( ( "State record updated. New state=%d.",
-                   publishRecordState ) );
+
+        if( status == MQTTSuccess )
+        {
+            LogInfo( ( "State record updated. New state=%s.",
+                       MQTT_State_strerror( publishRecordState ) ) );
+        }
+
+        /* A collision in the state engine and a duplicate flag set in the
+         * incoming packet would be indicating a duplicate incoming publish. */
+        else if( ( status == MQTTStateCollision ) && ( publishInfo.dup == true ) )
+        {
+            status = MQTTSuccess;
```

Comment:
If there was a state collision for an incoming publish, shouldn't we reset the state to PubRecSend/PubAckSend? Otherwise there will be more state collisions when we're sending the acks. It looks like if there was a collision, then publishRecordState will be MQTTStateNull, meaning sendPublishAcks on line 772 won't send any acks for it. However, according to the spec: 

"Until it has received the corresponding PUBREL packet, the Receiver MUST acknowledge any subsequent PUBLISH packet with the same Packet Identifier by sending a PUBREC. It MUST NOT cause duplicate messages to be delivered to any onward recipients in this case"

For QoS 2, I think we will always end up sending a PUBREC. The only state that occurs after receiving the PUBREL is PubCompSend. If we haven't sent the PUBCOMP, then the server has no way of knowing whether we ever received the PUBREL, so it should be fine the send the PUBREC. If we have sent the PUBCOMP, then we're required to treat this as a new publish and send a PUBREC.

Reply:
As discussed, the state returned from MQTT_UpdateStatePublish will either be PubRecSend/PubAckSend depending on the QoS level of the publish.  So that will take care of sending the acks.

Also if we have already sent a PUBCOMP, the state of the publish will not be present and this will be considered as a new publish.

Reply:
After offline conversation, I think there should be comment explaining each case, to avoid future confusion

Reply:
Agree.